File: Finite State Machines/fsm.c
#include <string.h>
#include <stdio.h>

typedef struct Transition
{
    char* init_state;
    char* combo_state;
    char* final_state;
} *transition;


transition new_transition(char* is,char* cs, char* fs)
{
    transition tor = (transition) malloc(sizeof(struct Transition));
    tor->init_state = is;
    tor->combo_state = cs;
    tor->final_state = fs;

    return tor;
}

void extract_fill_trans(char* string, int fields, transition* arr)
{
    char* start = string;
    char* end;
    int index = 0;

    //store initial state
    end = strchr(start, ':');
    char* init_state = (char*) malloc(256*sizeof(char));
    strncpy(init_state, start, end-start);
    start = end+1;

    //chop off initial state
    char* newstr = (char*) malloc(256*sizeof(char));
    strncpy(newstr, start, strlen(string)-strlen(newstr));

    for(int i=0; i<fields; i++)
    {
        end = strchr(newstr, ':');
        char* combo_state[256];
        strncpy(newstr, start, end-start);
        start = end+1;

        end = strchr(newstr, ' ');
        char* final_state[256];
        strncpy(newstr, start, end-start);
        start = end+1;
        end = strchr(newstr, ' ');

        arr[i] = new_transition(init_state, combo_state, final_state);

    }

    return;

}

int countFields(char* string)
{
    char * start = string;
    char * end;
    int count=1;
    while ( ( end = strchr( start, ',' ) ) != NULL )
    {
        start = end + 1;
        count++;
    }
    printf("Fields = %d\n", count);
    return count;
}


void fill_states(char* filename, char** arr)
{
    char* line = NULL;
    size_t len = 0;
    ssize_t read;
    FILE* fp;

    fp = fopen(filename, "r");
    int begin_read = 0;
    while((read = getline(&line, &len, fp))!=-1)
    {
        if(strstr(line, "[states]")!=NULL)
        {
            begin_read=1;
            continue;
        }
        else if(strstr(line, "[transitions]")!=NULL) begin_read = 0;
        else if(begin_read==0) continue;
        else if(strstr(line, "names:")!=NULL)
        {
            extract_fill_trans(line, countFields(line), arr);
            
        }

    }
}

void fill_transitions(char* filename, transition* arr)
{
    char* line = NULL;
    size_t len = 0;
    ssize_t read;
    FILE* fp;

    fp = fopen(filename, "r");
    int begin_read = 0;
    while((read = getline(&line, &len, fp))!=-1)
    {
        if(strstr(line, "[transitions]")!=NULL)
        {
            begin_read=1;
            continue;
        }
        else if(strstr(line, "[states]")!=NULL) begin_read = 0;
        else if(begin_read==0) continue;
        else if(strstr(line, ":")!=NULL)
        {
            extract_fill_trans(line, countFields(line), arr);
            
        }

    }
}

int main(int argc, char** argv)
{
    char* filename = argv[1];
    FILE* fp;
    char* line = NULL;
    size_t len = 0;
    ssize_t read;
    char c;


    transition* arr[256];
    char** states[256];
    fill_transitions(filename, arr);
    fill_states(filename, states);


    fp = fopen(filename, "r");
    char* state = states[0];
    while(c = fgetc(fp)!=EOF)
    {
        int match = 0;
        for(int i=0; i<256; i++)
        {
            if(states[0] == c)
            {
                int index = strchr(arr[i], ':');
                char* substr[256];
                strncpy(substr, arr[i], 1);
                state = substr;

            }
        }
    }

    if(state == states[0]) printf("Accept");
    else printf("Reject.");

    fclose(fp);
    if(line)
        free(line);

}

File: Finite State Machines/fsm.py
import sys
import re

#helper method to deterministic
def nodups(ls):
	#from first to 2nd to last element
	for i in range(0, len(ls)-1):
		for x in range(i+1, len(ls)):
			if(ls[i] == ls[x]):
				return 0
	return 1

def deterministic(ls):
	for sublist in ls:
		first_chars = []
		for item in sublist:
			first_chars.append(item[0])
		if(nodups(first_chars)==0): return 0
	return 1

#removes newline from last item in list
def clean(ls):
	item = ls[len(ls)-1]
	ls[len(ls)-1] = item[:len(item)-1]
	return ls;

def fill_states(fileadd):
	#makes assumption that start and end are first/second to last in names respectively
	fi = open(fileadd)
	li = []
	begin_read = 0
	for line in fi:
		if("[states]" in line):
			begin_read = 1
			continue
		elif("[transitions]" in line):
			begin_read = 0
			continue
		elif(begin_read == 0):
			continue
		elif("names:" in line):
			temp = re.split(r": *",line)
			els = re.split(r", *", temp[1])
			for item in els:
				li.append(item)
		# elif("end:" in line):
		# 	li = clean(li)
		# 	temp = re.split(r": *", line)
		# 	temp = clean(temp)
		# 	end = temp[1]
		# 	li.remove(end)
		# 	li.append(end)
		else:
			continue
	return li


#stores transitions in 2D array in format: 'a:1'
def fill_trans(fileadd):
	#assumes transitions listed in order of names
	fi = open(fileadd)
	trans = []
	begin_read = 0
	for line in fi:
		if("[transitions]" in line):
			begin_read=1
			continue
		elif(begin_read==0):
			continue
		elif("[states]" in line):
			begin_read = 0
			continue
		else:
			#extract name
			temp = line.split(":")
			name = temp[0]
			#get and store transitions
			temp2 = re.findall(r".:\w+",line)
			#temp2 = temp2[1:]
			trans.append(temp2);
	return trans;

fi = sys.argv[1]
states = clean(fill_states(fi))
trans = fill_trans(fi)
instr = sys.argv[2]

print("states=")
print(states)
print("Trans = ")
print(trans)
#print("determ = ",deterministic(trans))

if(deterministic(trans)==0):
	print("Error: nondeterministic function given.")
	sys.exit()

state = states[0]
for char in instr:
	#for each transition matching current state
	match = 0;
	for t in trans[states.index(state)]:
		#if exact match on state
		if(t[0] == char):
			temp = t.split(":")
			state = temp[1]
			match = 1
			break
	#if no match found, use wildcard
	if(match==0):
		for t in trans[states.index(state)]:
			if(t[0] == '*'):
				temp = t.split(":")
				state = temp[1]
				match = 1
				break
		#if cases were non-exhaustive
		if(match==0):
			print("Error: non-exhautive function given.")
			sys.exit()

if(state == states[len(states)-2]):
	print("Accept")
else:
	print("Reject. Ended in state", state)


